### ISSUE TITLE
Proposal: make `decodeError` public

### ISSUE THREAD

**Issue by kanata2** [jszwec/csvutil#48]
### Description
In v1.6 and later, it is possible to get more detailed errors.
However if you want to use information such as `Line` or `Column` field of `decodeError`,
(for example, when you want to customize error messages.(e.g. for i18n)), you cannot easily retrieve these.
How about making the `decodeError` public for such cases?

Or if you already have a case where you can handle `decodeError` properly, please let me know 🙏 

#### workaround for now
This is how we currently get it.

```go
var ute *csvutil.UnmarshalTypeError
if !errors.As(err, &ute) {
  // skip
}
var idx int
for i, c := range decoder.Record() {
  if c == ute.Value {
    idx = i
  }
}
line, col := csvReader.FieldPos(idx)
```

related: #43

**Comment by jszwec** [comment]
Hey, there is currently no way to handle `decodeError`. I had two reasons why not to expose decodeError immediately:

1. The original request for it seemed to seek better error message, over actually inspecting this information. That's why I didn't want to add another Error type to the package's API for as long as someone didn't need it
2. Line and Column fields are only available in Go1.17+ (due to `FieldPos`), which makes it a little awkward, because in older Go versions they will always be 0 making them useless

I think we can solve point 2 by adding proper comments on the Error type and warn the callers upfront about this, 
or define `DecodeError` for older Go versions without these fields

**Comment by kanata2** [comment]
Thanks for your feedback.

> The original request for it seemed to seek better error message, over actually inspecting this information.

Our use case is that we need to be able to distinguish whether an error is a decodeError or not,
and in that case we want to return a more informative error to the user.
In this usecase, how should users properly use csvutil?

**Comment by jszwec** [comment]
I think you misunderstood me. I agree with making it public. I only gave you the reason why it is unexported right now.

**Comment by kanata2** [comment]
Oh, sorry. I was misread...😢

**Comment by kanata2** [comment]
@jszwec Thanks❤️

**Comment by adamantike** [comment]
@jszwec, could this project have a new formal release, to be able to use this fix? Right now, the latest version `1.6.0` doesn't include this commit.

**Comment by jszwec** [comment]
@adamantike done
